File: hsg12-tinh/2024-2025-lamdong/tamgiac/tamgiac.py
import os
import sys

input_file = os.path.join(sys.path[0], 'tamgiac.inp')
output_file = os.path.join(sys.path[0], 'tamgiac.out')


class point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

A, B, C, M = point(0, 0), point(0, 0), point(0, 0), point(0, 0)

# Số epsilon
eps = 1e-9


def input_data():
    global A, B, C, M
    with open(input_file, 'r') as f:
        A.x, A.y = map(float, f.readline().strip().split())
        B.x, B.y = map(float, f.readline().strip().split())
        C.x, C.y = map(float, f.readline().strip().split())
        M.x, M.y = map(float, f.readline().strip().split())


# Hàm dùng để tính diện tích tam giác
def area(a, b, c):
    return abs((a.x * (b.y - c.y) + b.x * (c.y - a.y) + c.x * (a.y - b.y)) / 2)


def process():
    global A, B, C, M

    # Tính diện tích tam giác ABC
    abc = area(A, B, C)

    # Tính diện tích ba tam giác ABM, ACM, BCM
    abm = area(A, B, M)
    acm = area(A, C, M)
    bcm = area(B, C, M)

    # Nếu tổng diện tích ba tam giác con bằng diện tích tam giác ABC thì M thuộc tam giác ABC
    if abs(abc - abm - acm - bcm) < eps:
        result = 1
    else:
        result = 0

    with open(output_file, 'w') as f:
        f.write(str(result))


if __name__ == '__main__':
    input_data()
    process()

File: hsg12-tinh/2024-2025-lamdong/tamgiac/tamgiac.cpp
#include <iostream>

#define input_file "tamgiac.inp"
#define output_file "tamgiac.out"

using namespace std;

struct point
{
    int x, y;
};

point A, B, C, M;

// Số epsilon
const double eps = 1e-9;


void input()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    freopen(input_file, "r", stdin);

    cin >> A.x >> A.y;
    cin >> B.x >> B.y;
    cin >> C.x >> C.y;
    cin >> M.x >> M.y;
}


// Hàm dùng để tính diện tích tam giác
double area(point a, point b, point c)
{
    return abs((a.x * (b.y - c.y) + b.x * (c.y - a.y) + c.x * (a.y - b.y)) / 2.0);
}


void process()
{
    // Tính diện tích tam giác ABC
    double abc = area(A, B, C);

    // Tính diện tích ba tam giác ABM, ACM, BCM
    double abm = area(A, B, M);
    double acm = area(A, C, M);
    double bcm = area(B, C, M);

    freopen(output_file, "w", stdout);

    // Nếu tổng diện tích ba tam giác con bằng diện tích tam giác ABC thì M thuộc tam giác ABC
    if (abs(abc - abm - acm - bcm) < eps)
        cout << 1;
    else
        cout << 0;
}


int main()
{
    input();
    process();

    return 1;
}